Subbus/Classes/Subbus.swift

```diff
@@ -8,141 +8,241 @@
         
-        //Remove Listeners From Cache
-        while let index = shared.listeners.index(where: filter(identifier: identifier)) { //TODO: Maybe do structs; then you can probably do == comparison instead of filter as it stands
-            shared.listeners.remove(at: index)
-        }
-        
-        //Log
-        log("Unregistered all listeners for identifier \(identifier)")
+        log("Unegistered listener for identifier \"\(id)\"")
     }
     
-    fileprivate static func filter(identifier: String, eventName: String? = nil) -> (ListenerWrapper) -> Bool {
-        return {($0.identifier == identifier) && (eventName != nil ? $0.eventName == eventName : true)}
-    }
-    
-    //Misc
-    static func log(_ message: String) {
-        guard loggingEnabled == true else { return }
-        print("Subbus: \(message)")
+    public static func unsubscribe<T>(event: T.Type) {
+        let eventType = String(reflecting: T.self)
+        subscriptions.removeAll(where: { $0.matches(eventType: eventType) })
     }
-    
-    static func stringFor<I>(id: I) -> String? {
+}
+
+extension Subbus {
+    internal static func stringFor(id: Any) -> String? {
         var valString = "" //Value or pointer
-        
+        let idType = type(of: id)
+
         //Get the right string type; objects need memory address and primitives just need value.
-        if I.self is AnyClass {
+        if idType is AnyClass {
             valString = "\(Unmanaged.passUnretained(id as AnyObject).toOpaque())"
         } else {
             valString = "\(id)"
         }
-        
+
         //Ensure that the idString is valid
         guard valString.isEmpty == false else { return nil }
         guard valString != "nil" else { return nil }
+
+        return "\(idType)-\(valString)"
+    }
+    
+    internal static func log(_ message: String, force: Bool = false) {
+        guard logToConsole || force else { return }
+        print("Subbus: \(message)")
+    }
+    
+    internal static func cleanup() {
+        subscriptions.removeAll(where: { $0.identifier == nil })
+    }
+    
+    internal static func clearHistory() {
+        unhandledEvents = []
+        log("WARNING: cleared event history!  This function was made for unit testing, please don't use it.", force: true)
+    }
+    
+    internal static func clearSubscribers() {
+        subscriptions = []
+        log("WARNING: cleared subscriber cache!  This function was made for unit testing, please don't use it.", force: true)
+    }
+}
+
+internal struct Subscription {
+    weak var identifier: AnyObject?
+    var identifierString: String = ""
+    var eventType: String
+    var handler: Any
+    
+    init?(identifier: Any, eventType: String, handler: Any) {
+        self.eventType = eventType
+        self.handler = handler
+        
+        let (id, message) = parseId(identifier)
+        
+        guard let parsedId = id, message == nil else {
```

Comment:
I'd recommend moving this guard logic to the start of the function. I think it will help the init read more easily since we don't have a potential nil case in the middle.

Reply:
It actually kinda already is.  Being a struct, *all* variables need to be initialized before you can use any instance methods.  Of course optional and pre-defined variables don't need to be.  I have to pull `id` and `message` from the `parseId` function before I can know if I have a valid identifier or not.

Reply:
You could overcome this by making `parseId` a static method since it does not depend on a Subscription instance.